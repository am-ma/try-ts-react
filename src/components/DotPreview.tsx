import React, { CSSProperties } from 'react';
import DotValues from '~/models/DotValues.model';

interface DotPreviewProps {
  dots: DotValues[];
  dotSize: number;
  height: number;
  width: number;
}

interface DotPreviewState {
  dlUrl: string;
}

export default class DotPreview extends React.Component<DotPreviewProps, DotPreviewState> {
  private svg = React.createRef<SVGSVGElement>();
  private canvas = React.createRef<HTMLCanvasElement>();
  constructor(props: DotPreviewProps) {
    super(props);
    this.state = {
      dlUrl: '#',
    };
  }
  render() {
    // render svg
    const dotSize = this.props.dotSize;
    const width = dotSize * this.props.width;
    const height = dotSize * this.props.height;

    const imageDots = this.props.dots
      .map((dot, i) => {
        if (!dot.isFill) {
          return null;
        }
        const style: CSSProperties = {
          fill: dot.color,
        };
        const xIndex = i % this.props.height;
        const yIndex = Math.floor(i / this.props.width);
        return (
          <rect
            key={`${xIndex}_${yIndex}`}
            x={xIndex * dotSize}
            y={yIndex * dotSize}
            width={dotSize}
            height={dotSize}
            style={style}></rect>
        );
      })
      .filter(dotJSX => dotJSX !== null);

    return (
      <>
        <svg id="svg" width={width} height={height} ref={this.svg}>
          {imageDots}
        </svg>
        <canvas id="canvas" ref={this.canvas} width={width} height={height} className="not-display"></canvas>
        <div>
          <a href={this.state.dlUrl} id="download" download="dots.png">
            download
          </a>
        </div>
      </>
    );
  }

  componentDidMount() {
    this.renderForDownload();
  }

  componentDidUpdate() {
    this.renderForDownload();
  }

  renderForDownload() {
    // svg to xml data src
    const svg = this.svg.current!;
    const svgData = new XMLSerializer().serializeToString(svg as Node);
    const imgsrc = 'data:image/svg+xml;charset=utf-8;base64,' + btoa(unescape(encodeURIComponent(svgData)));

    const canvas = this.canvas.current!;
    const ctx = canvas.getContext('2d') as CanvasRenderingContext2D;
    const image = new Image();
    image.onload = () => {
      // render canvas
      ctx.drawImage(image, 0, 0);
      // render download button
      this.setState({
        dlUrl: canvas.toDataURL('image/png'),
      });
    };
    image.src = imgsrc;
  }
}
