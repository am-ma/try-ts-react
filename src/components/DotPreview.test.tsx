import React from 'react';
import DotPreview from './DotPreview';
import DotValues from '~/models/DotValues.model';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

describe('DotPreview', () => {
  const dot1: DotValues = { isFill: true, color: '#000' };
  const dot2: DotValues = { isFill: true, color: '#000' };
  const dot3: DotValues = { isFill: false, color: '#000' };
  const dot4: DotValues = { isFill: false, color: '#000' };
  const dots = [dot1, dot2, dot3, dot4];
  const jsx = <DotPreview dots={dots} dotSize={8} height={2} width={2} />;
  describe('render', () => {
    it('will render', () => {
      jest.spyOn(DotPreview.prototype, 'renderForDownload').mockImplementationOnce(jest.fn());

      const result = renderer.create(jsx).toJSON();

      expect(result).toMatchSnapshot();
    });
  });

  describe('componentDidMount', () => {
    it('will call renderForDownload()', () => {
      jest.spyOn(DotPreview.prototype, 'renderForDownload').mockImplementationOnce(jest.fn());

      const wrapper = shallow(jsx);
      const dotPreview = wrapper.instance() as DotPreview;

      expect(dotPreview.renderForDownload).toHaveBeenCalled();
    });
  });

  describe('componentDidUpdate', () => {
    it('will call renderForDownload()', () => {
      jest.spyOn(DotPreview.prototype, 'renderForDownload').mockImplementationOnce(jest.fn());

      const wrapper = shallow(jsx);
      const dotPreview = wrapper.instance() as DotPreview;

      const dot1: DotValues = { isFill: true, color: '#FFF' };
      const dot2: DotValues = { isFill: true, color: '#FFF' };
      const dot3: DotValues = { isFill: false, color: '#000' };
      const dot4: DotValues = { isFill: false, color: '#000' };
      const newDots = [dot1, dot2, dot3, dot4];

      dotPreview.renderForDownload = jest.fn();
      wrapper.setState({ dots: newDots });

      expect(dotPreview.renderForDownload).toHaveBeenCalled();
    });
  });
});
