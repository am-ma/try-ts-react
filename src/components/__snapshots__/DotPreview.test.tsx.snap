// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DotPreview render will render 1`] = `
Array [
  <svg
    height={16}
    id="svg"
    width={16}
  >
    <rect
      height={8}
      style={
        Object {
          "fill": "#000",
        }
      }
      width={8}
      x={0}
      y={0}
    />
    <rect
      height={8}
      style={
        Object {
          "fill": "#000",
        }
      }
      width={8}
      x={8}
      y={0}
    />
  </svg>,
  <canvas
    className="not-display"
    height={16}
    id="canvas"
    width={16}
  />,
  <div>
    <a
      download="dots.png"
      href="#"
      id="download"
    >
      download
    </a>
  </div>,
]
`;
